WeakMap.ts

```diff
@@ -0,0 +1,62 @@
+import core = require('./interfaces');
+import has = require('./has');
+import lang = require('./lang');
+
+has.add('es6-weak-map', typeof (<any> WeakMap) !== 'undefined');
+
+class WeakMapPolyfill<K, V> implements core.IWeakMap<K, V> {
+	private _name:string = undefined;
+	static length:number = 1;
+
+	constructor(iterable:any) {
+		this._name = '__wm' + lang.getUID() + (startId++ + '__');
+		if (iterable && 'forEach' in iterable) {
+			iterable.forEach((item:any, i:number) => {
+				if (Array.isArray(item) && item.length === 2) {
+					this.set(iterable[i][0], iterable[i][1]);
+				}
+				else {
+					this.set(iterable[i], i);
+				}
+			});
+		}
+	}
+
+	set(key:any, value:any):any {
+		var entry = key[this._name];
+		if (entry && entry[0] === key) {
+			entry[1] = value;
+		}
+		else {
+			Object.defineProperty(key, this._name, {
+				value: [key, value],
+				writable: true
+			});
+		}
+		return this;
+	}
+
+	get(key:any):any {
+		var entry = key[this._name];
+		return entry && entry[0] === key ? entry[1] : undefined;
+	}
+
+	has(key:any):boolean {
+		var entry = key[this._name];
+		return Boolean(entry && entry[0] === key && entry[1]);
+	}
+
+	delete(key:any):void {
+		this.set(key, undefined);
+	}
+}
+
+if (!has('es6-weak-map')) {
+	var startId:number = Math.floor(Date.now() % 100000000);
+}
+
+var WM:any;
```

Comment:
It just seems impossible to annotate this in a way that it compiles in TypeScript.  Trying to specify `var WM:WeakMap|WeakMapPolyfill` results in a TS2314.  It seems that at the moment this is wholly intentional, because it is believed to be "useless" as per Microsoft/TypeScript#2042.